<a href="https://colab.research.google.com/github/DavidRQ24/tests-python-/blob/main/Modulo%201/Copia_de_Introducci%C3%B3n_a_Funciones.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img src="https://www.ctic.uni.edu.pe/wp-content/uploads/2022/04/588px-x-348px-web-1.png" alt="HTML5 Icon" width="900" height="350" >

# **Introducción a Funciones**

---

**Objetivo:** Al finalizar el laboratorio, los estudiantes tendrás  conocimiento sobre la creación y uso de funciones en Python, incluyendo parámetros, valores de retorno, alcance de variables, funciones lambda, recursión, manejo de excepciones y documentación de funciones.

## Tabla de Contenidos

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>

1. <a href="#item31">Introducción a las Funciones</a>
2. <a href="#item31">Parámetros y Argumentos</a>
3. <a href="#item31">Valores de Retorno</a>
4. <a href="#item31">Alcance de Variables (Scope)</a>
5. <a href="#item31">Funciones Lambda</a>
6. <a href="#item31">Recursión</a>
7. <a href="#item31">Manejo de Excepciones en Funciones</a>
8. <a href="#item31">Documentación de Funciones</a>


</font>
</div>

---

## 1. Introducción a las Funciones

### Descripción:



**Definición y sintaxis básica:**

Una función es un bloque de código reutilizable que realiza una tarea específica.


Sintaxis básica:

In [ ]:
def nombre_funcion(parámetros):
    """Docstring opcional."""
    # Cuerpo de la función
    return valor_de_retorno


In [5]:
def greetings(name, course):
  '''Esta función genera un saludo
  para los estudiantes
  ---parametros---
  name: el nombre de la persona a saludar
  course: nombre del curso del estudiante'''
  print(f"¡Hola {name}! Bienvenido a la clase de {course}")

In [6]:
greetings(name = 'David', course = 'Python')

¡Hola David! Bienvenido a la clase de Python


In [7]:
greetings('David', 'Python')

¡Hola David! Bienvenido a la clase de Python


In [8]:
greetings(course = 'Python', name = 'David') #si se pone al reves sin los valores no se interpreta y lo pone como tal

¡Hola David! Bienvenido a la clase de Python


In [3]:
greetings? #Obtienes la información de la documentación de la función

In [9]:
def greetings2(name, course = 'Python'):  #Cuando no quiero que se modifique todos los parametros, pero si puedo cambiarlas luego, siempre se colocan a la derecha los seteados
  '''Esta función genera un saludo
  para los estudiantes
  ---parametros---
  name: el nombre de la persona a saludar
  course: nombre del curso del estudiante'''
  print(f"¡Hola {name}! Bienvenido a la clase de {course}")

In [13]:
greetings2(name = 'David') #Ya no da error al no ingresar el course porque ya lo tengo seteado

¡Hola David! Bienvenido a la clase de Python


In [14]:
greetings2(name = 'David', course = 'C++') #Puedo cambiar el curso seteado

¡Hola David! Bienvenido a la clase de C++


In [22]:
from math import exp

def sigmoid(x):
    '''La función genera una transformación sigmoide del input x'''
    return 1/(1+exp(-x)) #si se deja solo con un print o sin el print no devuelve un numero como tal, por eso es importante el uso del return

In [23]:
sigmoid(2)*2

1.7615941559557646

In [24]:
from math import exp

def mix_functions(p, x):
    if p < 0.5:
        return 1/(1+exp(-x))
    else:
        return 1/(1+exp(x))


**Ventajas de usar funciones:**

- Reutilización de código.
- Modularidad y organización.
- Facilita el mantenimiento y la lectura del código.

###Ejemplos:



**Ejemplo 1.1:** Función sin parámetros:


In [ ]:
def saludo():
    """Función que imprime un saludo."""
    print("¡Hola, bienvenidos al curso de Python!")

saludo()  # Llamada a la función


**Ejemplo 1.2:** Función con parámetros:

In [ ]:
def saludo_personalizado(nombre):
    """Función que saluda a una persona específica."""
    print(f"¡Hola, {nombre}! Bienvenido al curso de Python.")

saludo_personalizado("Carlos")


**Ejemplo 1.3:** Función que calcula el área de un círculo:

In [ ]:
import math

def area_circulo(radio):
    """Calcula el área de un círculo dado su radio."""
    return math.pi * radio ** 2

print(area_circulo(5))  # Imprime el área del círculo con radio 5


**Ejemplo 1.4:** Función que determina si un número es par o impar:

In [ ]:
def es_par(n):
    """Verifica si un número es par o impar."""
    if n % 2 == 0:
        return True
    else:
        return False

print(es_par(10))  # Imprime True
print(es_par(7))   # Imprime False


**Ejemplo 1.5** Función que genera una lista de números aleatorios:


In [ ]:
import random

def generar_numeros_aleatorios(cantidad, inicio, fin):
    """Genera una lista de números aleatorios."""
    numeros = []
    for _ in range(cantidad):
        numeros.append(random.randint(inicio, fin))
    return numeros

print(generar_numeros_aleatorios(5, 1, 100))


### Ejercicios:


**Ejercicio 1.1.**

Cálculo de distancia euclidiana:

Crea una función distancia_euclidiana(punto1, punto2) que calcule la distancia entre dos puntos en un plano 2D.
Tip: Los puntos pueden representarse como tuplas (x, y).



In [28]:

def distancia_euclidiana(punto1, punto2):
    """Calcula la distancia euclidiana entre dos puntos en un plano 2D.
    ----parametros----
    punto1: tupla de coordenada x1, y1
    punto2: tupla de coordenada x2, y2"""
    distancia = ((punto1[0] - punto2[0])**2 + (punto1[1] - punto2[1])**2)**0.5
    return distancia

distancia_euclidiana((2,3), (4,7))


4.47213595499958

**Ejercicio 1.2.**

Conversión de números romanos:

Escribe una función romano_a_entero(romano) que convierta un número romano a un entero.


In [46]:
def romano_a_entero(romano):
    """Convierte un romano a entero
    ----parametros----
    romano: string con el numero romano"""
    romanos = {'I': 1, 'V': 5, 'X': 10, 'L': 50, 'C': 100, 'D': 500, 'M': 1000}
    entero = 0
    for i in range(len(romano)):
        if i > 0 and romanos[romano[i]] > romanos[romano[i - 1]]:
            entero += romanos[romano[i]] - 2 * romanos[romano[i - 1]]
        else:
            entero += romanos[romano[i]]
    return entero

romano_a_entero('XC')

9

**Ejercicio 1.3**

Generador de contraseñas seguras:

Define una función generar_contraseña(longitud) que genere una contraseña aleatoria que incluya letras, números y símbolos.
Tip: Usa el módulo random y la biblioteca string.

In [39]:
from random import choice
import string


def generar_contraseña(longitud):
    '''Genera una contraseña aleatoria que incluya letras, números y símbolos.
    ----parametros----
    longitud: longitud de la contraseña'''

    cadena = string.digits + string.punctuation + string.ascii_letters
    contraseña = ''

    for i in range(longitud):           #Otra forma con while len(contraseña) < longitud:
        contraseña += choice(cadena)    #contraseña = contraseña + choice(caracteres)
    return contraseña


In [49]:
generar_contraseña(8)

'exHzh;Ip'

**Ejercicio 1.4**

Verificación de palíndromos:

Crea una función es_palindromo(cadena) que verifique si una cadena es un palíndromo (se lee igual al derecho y al revés).
Desafío: Ignora espacios y considera mayúsculas y minúsculas.

In [56]:
def es_palindromo(cadena):
    '''Verifica si un string es palindromo o no
    ----parametros----
    cadena: string a verificar'''
    cadena = cadena.lower().rstrip()

    if cadena == cadena[::-1]:
        return True
    else:
        return False

In [59]:
es_palindromo('Reconocer')

True

**Ejercicio 1.5**

Análisis de frecuencia de palabras:

Escribe una función frecuencia_palabras(texto) que reciba un texto y devuelva un diccionario con la frecuencia de cada palabra.
Tip: Usa el método split() y considera eliminar puntuación.


In [74]:
import re

def frecuencia_palabras(texto):
    '''Diccionario con la frecuencia de cada palabra
    ----parametros----
    texto: string a verificar'''
    texto_new = re.sub(r'[^\w\s]', '', texto)
    texto_new = texto_new.split()

    frecuencia = {}
    for palabra in texto_new:
        if palabra in frecuencia:
            frecuencia[palabra] += 1
        else:
            frecuencia[palabra] = 1
    return frecuencia


In [75]:
frecuencia_palabras('este texto no se repite porque es un texto que no se repite si no es se')

{'este': 1,
 'texto': 2,
 'no': 3,
 'se': 3,
 'repite': 2,
 'porque': 1,
 'es': 2,
 'un': 1,
 'que': 1,
 'si': 1}

**Ejercicio 1.6**

Conversión de unidades de temperatura:

Define una función convertir_temperatura(valor, origen, destino) que convierta temperaturas entre Celsius, Fahrenheit y Kelvin.
Desafío: Maneja casos inválidos y verifica las unidades de entrada.


In [81]:
def convertir_temperatura(valor, origen, destino):
    '''Convertor de temperaturas entre celsius, fahrenheit y kelvin
    ----parametros----
    valor: temperatura origen
    origen: unidad de medida de temperatura origen (C, F, K)
    destino: unidad de medida de temperatura destino (C, F, K)'''
    #Formula 5 * (F - 32) = 9 * (C) y K = C + 273.15

    origen.upper()
    destino.upper()

    if origen == 'C' and destino == 'K':
        valor = valor + 273.15
        return valor
    elif origen == 'C' and destino == 'F':
          valor =  (9/5) * valor + 32
          return valor
    elif origen == 'F' and destino == 'K':
          valor = (5/9) * (valor - 32) + 273.15
          return valor
    elif origen == 'F' and destino == 'C':
          valor = (5/9) * (valor - 32)
          return valor
    elif origen == 'K' and destino == 'C':
          valor = valor - 273.15
          return valor
    elif origen == 'K' and destino == 'F':
          valor = (9/5) * (valor - 273.15) + 32
          return valor
    else:
        return(valor)

In [82]:
convertir_temperatura(37, 'F', 'K')

275.92777777777775

**Ejercicio 1.7**

Calculadora de propinas:

Crea una función calcular_propina(total, porcentaje=10) que calcule la propina a partir del total de una cuenta.
Tip: El parámetro porcentaje debe ser opcional.


**Ejercicio 1.8**

Determinación de años bisiestos:

Escribe una función es_bisiesto(año) que determine si un año es bisiesto.
Desafío: Considera las reglas de los años divisibles por 100 y 400.


**Ejercicio 1.9**

Generador de números Fibonacci:


Define una función generar_fibonacci(n) que retorne una lista con los primeros n números de la serie de Fibonacci.
Tip: Usa un bucle para generar la serie.





**Ejercicio 1.10**

Calculadora de índice de masa corporal (IMC):

Crea una función calcular_imc(peso, altura) que calcule el IMC y retorne una categoría (bajo peso, normal, sobrepeso, obesidad).
Desafío: Usa las categorías estándar de la OMS.

**Ejercicio 1.11**

Juego de adivinanza:

Escribe una función adivinar_numero() que genere un número aleatorio y permita al usuario adivinarlo, indicando si el número ingresado es mayor o menor hasta que acierte.

## 2. Parámetros y Argumentos

**Descripción:**

**Parámetros posicionales y argumentos:**

Los parámetros se definen en la declaración de la función y los argumentos se pasan al llamar a la función.

**Parámetros con valores por defecto:**

- Permiten que ciertos parámetros sean opcionales.
- Argumentos nombrados (keyword arguments):
- Especificar argumentos por nombre, independientemente del orden.

**Argumentos arbitrarios:**

- *args: Permite pasar un número variable de argumentos posicionales.

- **kwargs: Permite pasar un número variable de argumentos nombrados.

### Ejemplos:


**Ejemplo 2.1.:** Función con parámetros por defecto y argumentos nombrados:


In [ ]:
def enviar_email(destinatario, asunto="Sin asunto", mensaje=""):
    """Envía un email simulado con los detalles proporcionados."""
    print(f"Enviando email a: {destinatario}")
    print(f"Asunto: {asunto}")
    print(f"Mensaje: {mensaje}")

enviar_email("usuario@example.com", mensaje="Hola, este es un mensaje de prueba.")


**Ejemplo 2.2.:** Uso de *args para sumar números:

In [ ]:
def sumar_numeros(*numeros):
    """Suma una cantidad variable de números."""
    total = sum(numeros)
    print(f"La suma de {numeros} es {total}")

sumar_numeros(1, 2, 3)
sumar_numeros(4, 5, 6, 7)


**Ejercicio 2.3:** Uso de **kwargs para crear un perfil de usuario:


In [ ]:
def crear_perfil(**informacion):
    """Crea un perfil de usuario con la información proporcionada."""
    print("Perfil de usuario:")
    for clave, valor in informacion.items():
        print(f"{clave.capitalize()}: {valor}")

crear_perfil(nombre="Laura", edad=28, ciudad="Barcelona", profesion="Ingeniera")


### Ejercicios:

**Ejercicio 2.1.**

Función de envío de notificaciones:

Escribe una función enviar_notificacion(mensaje, *usuarios) que envíe un mensaje a múltiples usuarios.


**Ejercicio 2.2.**

Gestor de pedidos:

Crea una función procesar_pedido(cliente, **articulos) que procese un pedido con diferentes artículos y sus cantidades.

**Ejercicio 2.3.**

Calculadora de estadísticas:

Define una función calcular_estadisticas(*datos) que devuelva la media, mediana y moda de una serie de números.




**Ejercicio 2.4.**

Generador de informes:

Escribe una función generar_informe(titulo, *secciones, **detalles) que genere un informe estructurado.



**Ejercicio 2.5.**

Simulador de votación:

Crea una función votar(opcion, **votantes) que registre los votos de diferentes personas.


**Ejercicio 2.6.**

Función de traducción:

Define una función traducir(texto, idioma_destino="inglés") que traduzca el texto al idioma especificado (simulado).




**Ejercicio 2.7.**

Personalizador de saludos:

Escribe una función saludar_personalizado(nombre, mensaje="Hola", **estilos) que salude aplicando diferentes estilos de texto (por ejemplo, mayúsculas, subrayado).

**Ejercicio 2.8.**

Creador de menús dinámicos:

 Crea una función crear_menu(titulo, **opciones) que genere un menú interactivo en consola.



**Ejercicio 2.9.**

Analizador de parámetros:

 Define una función analizar_parametros(*args, **kwargs) que imprima los tipos y valores de los argumentos recibidos.










**Ejercicio 2.10.**

Función matemática genérica:

 Escribe una función operar_numeros(operacion, *numeros) que aplique una operación matemática (suma, multiplicación) a una serie de números.

## 3. Valores de Retorno


**Descripción:**

**Uso de la palabra clave return:**
- Termina la ejecución de la función y opcionalmente devuelve un valor.

**Retorno de múltiples valores:**
- Las funciones pueden devolver múltiples valores utilizando tuplas.

**Retorno de funciones y objetos:**
- Las funciones pueden devolver otras funciones u objetos complejos.

### Ejemplos:

**Ejemplo 3.1.** Función que devuelve un diccionario:


In [ ]:
def analizar_cadena(cadena):
    """Devuelve un diccionario con la longitud y número de vocales de la cadena."""
    vocales = 'aeiouAEIOU'
    num_vocales = sum(1 for letra in cadena if letra in vocales)
    resultado = {
        'longitud': len(cadena),
        'vocales': num_vocales
    }
    return resultado

info = analizar_cadena("Hola Mundo")
print(info)


**Ejemplo 3.2** Función que retorna una función personalizada:

In [ ]:
def crear_funcion_operador(operador):
    """Crea una función que aplica el operador dado."""
    if operador == 'suma':
        def suma(a, b):
            return a + b
        return suma
    elif operador == 'resta':
        def resta(a, b):
            return a - b
        return resta
    else:
        return None

operacion_suma = crear_funcion_operador('suma')
resultado = operacion_suma(5, 3)
print(f"El resultado de la suma es {resultado}")


**Ejemplo 3.3** Función generador:


In [ ]:
def generador_pares(limite):
    """Genera números pares hasta el límite especificado."""
    numero = 0
    while numero <= limite:
        print(numero)
        numero += 2

generador_pares(10)

### Ejercicios:

**Ejercicio 3.1:** Desglosador de billetes: Crea una función desglosar_cantidad(cantidad) que devuelva el número mínimo de billetes y monedas necesarios para una cantidad dada.


**Ejercicio 3.2:**
Analizador de URLs: Escribe una función analizar_url(url) que retorne el protocolo, dominio y ruta de una URL.

**Ejercicio 3.3:**
Calculadora de estadísticas de texto: Define una función estadisticas_texto(texto) que devuelva el número de oraciones, palabras y caracteres.

**Ejercicio 3.4:**
Gestor de inventario: Crea una función actualizar_inventario(producto, cantidad, inventario) que actualice un diccionario de inventario y lo retorne.


**Ejercicio 3.5:** Generador de números primos: Escribe una función generador_primos() que retorne una función que genere números primos consecutivos cada vez que se llame.

**Ejercicio 3.6:**
Conversor de moneda con tasas variables: Define una función convertir_moneda(cantidad, moneda_origen, moneda_destino, tasas_cambio) que utilice un diccionario de tasas de cambio.

**Ejercicio 3.7:**
Calculador de ruta más corta: Crea una función ruta_mas_corta(grafo, inicio, fin) que devuelva la ruta más corta entre dos nodos (puedes simular el grafo como un diccionario).


**Ejercicio 3.8:**
Simulador de máquina expendedora: Escribe una función seleccionar_producto(producto, **inventario) que devuelva el precio y actualice el inventario.


**Ejercicio 3.9:**
 Sistema de recomendaciones: Define una función recomendar(producto, **productos_relacionados) que devuelva una lista de productos recomendados.

**Ejercicio 3.10:**
Análisis de sentimientos: Crea una función analizar_sentimiento(texto) que devuelva una puntuación de sentimiento (puedes simular el análisis).

## 4. Manejo de Excepciones en Funciones


**Descripción:**

**Uso de try, except, finally:**

- Manejo de errores y excepciones para evitar que el programa se detenga abruptamente.


**Propagación de excepciones:**

- Las excepciones pueden propagarse hacia arriba en la pila de llamadas si no son manejadas.

**Creación de excepciones personalizadas:**

- Definir excepciones propias para casos específicos.

### Ejemplos:

**Ejemplo 4.1.** Manejo de excepciones al convertir tipos:

In [ ]:
def convertir_a_entero(valor):
    """Convierte un valor a entero si es posible."""
    try:
        return int(valor)
    except (ValueError, TypeError):
        print("No se puede convertir el valor a entero.")
        return None

numero = convertir_a_entero("abc")


**Ejemplo 4.2.** Propagación de excepciones:

In [ ]:
def funcion_interna():
    raise ValueError("Ocurrió un error en la función interna.")

def funcion_externa():
    try:
        funcion_interna()
    except ValueError as e:
        print(f"Excepción capturada en función externa: {e}")

funcion_externa()


**Ejemplo 4.3.** Excepción personalizada para validación:

In [ ]:
def verificar_edad(edad):
    if edad < 0:
        raise ValueError("La edad no puede ser negativa.")
    else:
        print(f"La edad es {edad}")

try:
    verificar_edad(-5)
except ValueError as e:
    print(e)



La función raise en Python se usa para lanzar o "levantar" una excepción de manera explícita en el flujo de un programa. Esto es útil para manejar errores o situaciones específicas que requieran interrumpir la ejecución y notificar de un problema.

In [ ]:
def verificar_positivo(x):
    if x <= 0:
        raise ValueError("El valor debe ser positivo")
    return x

try:
    verificar_positivo(-4)
except ValueError as e:
    print(e)


In [ ]:
try:
    # Intentar algo que podría causar un error
    resultado = 10 / 0
except ZeroDivisionError as e:
    print("Se ha capturado una división por cero")


Se ha capturado una división por cero


- FileNotFoundError: Común al trabajar con archivos, cuando intentas abrir uno que no existe.

- ZeroDivisionError: Ocurre cuando intentas dividir un número por cero, algo frecuente en cálculos matemáticos.

- ValueError: Se lanza cuando un valor tiene el tipo correcto pero no el contenido adecuado (por ejemplo, al intentar convertir int("texto")).

- TypeError: Común cuando realizas operaciones en tipos incompatibles, como sumar una cadena y un número ("texto" + 5).

- IndexError: Sucede al intentar acceder a un índice fuera del rango en una lista o secuencia (lista[10] en una lista de solo 5 elementos).

- KeyError: Frecuente al trabajar con diccionarios cuando intentas acceder a una clave que no existe (mi_diccionario["clave_inexistente"]).

- AttributeError: Aparece al intentar acceder a un atributo o método que no existe en un objeto (cadena.inexistente() si cadena es un string sin ese método).

- NameError: Común si intentas usar una variable o función que no ha sido definida (por ejemplo, un error tipográfico en el nombre de una variable).

- ImportError / ModuleNotFoundError: Aparece al importar módulos que no existen o no están instalados.

- SyntaxError: Ocurre cuando hay errores de sintaxis en el código, como faltas de paréntesis o dos puntos en estructuras condicionales o bucles.

- IndentationError: Común en Python, donde la indentación es crucial; aparece cuando la indentación es incorrecta en el código.

**Finally**

El bloque finally en Python se usa junto con try y except para asegurar que un conjunto específico de instrucciones se ejecute siempre, sin importar si se lanzó una excepción o no. Es particularmente útil para liberar recursos, cerrar archivos, o realizar cualquier acción que deba ejecutarse sin falta, incluso si ocurre un error.

**Estructura básica**

In [ ]:
try:
    # Código que podría lanzar una excepción
    x = 10 / 0
except ZeroDivisionError:
    print("Se ha capturado una división por cero")
finally:
    print("Este bloque se ejecuta siempre, ocurra o no una excepción.")


In [ ]:
try:
    file = open("archivo.txt", "r")
    # Código para leer el archivo
except FileNotFoundError:
    print("El archivo no se encontró.")
finally:
    file.close()  # Se ejecuta siempre, asegurando que el archivo se cierre
    print("Archivo cerrado.")


### Ejercicios:

**Ejercicio 4.1:** Validador de entradas de usuario: Escribe una función que solicite un número al usuario y maneje excepciones por entradas inválidas.


**Ejercicio 4.2:** Lectura de archivos con manejo de errores: Crea una función que abra y lea un archivo, manejando posibles excepciones.


**Ejercicio 4.3:**  Sistema de autenticación: Implementa una función que valide credenciales de usuario y lance excepciones personalizadas en caso de errores.

**Ejercicio 4.4:** Cálculo con validación de datos: Escribe una función que realice operaciones matemáticas y maneje excepciones por divisiones entre cero o tipos de datos incorrectos.

**Ejercicio 4.5:**
Registro de errores en un sistema: Define una función que registre los errores ocurridos en un archivo de log.

**Ejercicio 4.6:**
Reintento automático después de una excepción: Crea una función que reintente una operación un número determinado de veces en caso de fallo.

**Ejercicio 4.7:**
Manejo de múltiples excepciones en funciones anidadas: Escribe funciones anidadas que manejen y propaguen diferentes excepciones.

**Ejercicio 4.8:**
Validación de formularios: Implementa una función que valide datos de un formulario y maneje excepciones por campos faltantes o inválidos.


**Ejercicio 4.9:**
Operaciones con recursos externos: Escribe una función que maneje excepciones al conectar con una base de datos simulada.

---

# Gracias por completar este laboratorio!

---
